### PR TITLE
Add option to retrieve oauth2.0 token using basic authentication

### PR DESCRIPTION
**Pull request by grahamrp**
Regarding issue #288 this adds an option to `oauth2.0_token` so that when the authorization code is exchanged for the token it can be done using http basic authentication, rather than the default method of including the app key and secret in the request body.

From the [OAuth2.0 Framework RFC](https://tools.ietf.org/html/rfc6749) it looks like authorization servers should provide the http basic authentication method as a minimum, and the method of providing the app key/secret in the request body is optional. I see that the github api allows both, but fitbit only allows basic authentication.

### CODE REVIEW

File: R/oauth-init.R [hunk 1]
```diff
@@ -51,11 +51,16 @@ init_oauth1.0 <- function(endpoint, app, permission = NULL,
...
+#'     retrieve the token. Some authorization servers require this.
+#'     If \code{FALSE}, the default, retrieve the token by including the
+#'     app key and secret in the request body.
 #' @export
 #' @keywords internal
 init_oauth2.0 <- function(endpoint, app, scope = NULL, type = NULL,
                           use_oob = getOption("httr_oob_default"),
-                          is_interactive = interactive()) {
+                          is_interactive = interactive(),
+                          use_basic_auth = NULL) {
```
Comment (hadley) [inline R/oauth-init.R]: It would be more clear to make the default value `FALSE`

File: R/oauth-init.R [hunk 2]
```diff
@@ -85,13 +90,23 @@ init_oauth2.0 <- function(endpoint, app, scope = NULL, type = NULL,
...
 
   # Use authorisation code to get (temporary) access token
-  req <- POST(endpoint$access, encode = "form",
-    body = list(
-      client_id = app$key,
-      client_secret = app$secret,
-      redirect_uri = redirect_uri,
-      grant_type = "authorization_code",
-      code = code))
+  if (isTRUE(use_basic_auth)){
```
Comment (hadley) [inline R/oauth-init.R]: Could you add a comment with link to the relevant section of the RFC?

File: R/oauth-token.r [hunk 3]
```diff
@@ -181,8 +181,9 @@ Token1.0 <- R6::R6Class("Token1.0", inherit = Token, list(
 #' caching policies used to store credentials across sessions.
 #'
 #' @inheritParams init_oauth2.0
-#' @param as_header If \code{TRUE}, the default, sends oauth in bearer header.
-#'   If \code{FALSE}, adds as parameter to url.
+#' @param as_header If \code{TRUE}, the default, configures the token to add
+#' itself to the bearer header of subsequent requests.
```
Comment (hadley) [inline R/oauth-token.r]: Missing indent

**Comment by hadley** [comment]
Looks good, apart from a few minor comments.  Could you please also a bullet to NEWS?

File: R/oauth-init.R [hunk 4]
```diff
@@ -85,13 +90,25 @@ init_oauth2.0 <- function(endpoint, app, scope = NULL, type = NULL,
...
-  req <- POST(endpoint$access, encode = "form",
-    body = list(
-      client_id = app$key,
-      client_secret = app$secret,
-      redirect_uri = redirect_uri,
-      grant_type = "authorization_code",
-      code = code))
+  # Send credentials using HTTP Basic or as parameters in the request body
+  # See https://tools.ietf.org/html/rfc6749#section-2.3 (Client Authentication)
+  if (isTRUE(use_basic_auth)){
```
Comment (hadley) [inline R/oauth-init.R]: It feels like there's a little too much duplication here. Could you please pull out the common `body`?

**Comment by grahamrp** [comment]
I've pulled out the common body now, thanks. I've also moved the NEWS bullet to the top. I hope the commits are okay - I'm a novice at collaboration on github.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**